src/Catalyst.Core.Modules.Dfs/Migration/MigrationManager.cs

```diff
@@ -0,0 +1,147 @@
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Catalyst.Abstractions.Dfs.Migration;
+using Catalyst.Abstractions.Options;
+using Common.Logging;
+
+namespace Catalyst.Core.Modules.Dfs.Migration
+{
+    /// <summary>
+    ///   Allows migration of the repository. 
+    /// </summary>
+    public sealed class MigrationManager : IMigrationManager
+    {
+        private readonly RepositoryOptions _options;
+        private static readonly ILog Log = LogManager.GetLogger(typeof(MigrationManager));
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="MigrationManager"/> class
+        ///   for the specifed <see cref="DfsService"/>.
+        /// </summary>
+        public MigrationManager(RepositoryOptions options)
+        {
+            _options = options;
+            Migrations = typeof(MigrationManager).Assembly.GetTypes()
+               .Where(x => typeof(IMigration).IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract)
+               .Select(x => (IMigration) Activator.CreateInstance(x))
+               .OrderBy(x => x.Version)
+               .ToList();
+        }
+
+        /// <summary>
+        ///   The list of migrations that can be performed.
+        /// </summary>
+        public List<IMigration> Migrations { get; }
+
+        /// <summary>
+        ///   Gets the latest supported version number of a repository.
+        /// </summary>
+        public int LatestVersion => Migrations.Last().Version;
+
+        /// <summary>
+        ///   Gets the current vesion number of the  repository.
+        /// </summary>
+        public int CurrentVersion
+        {
+            get
+            {
+                var path = VersionPath();
+                if (!File.Exists(path))
+                {
+                    return 0;
+                }
+                
+                using (var reader = new StreamReader(path))
+                {
+                    var s = reader.ReadLine();
+                    return int.Parse(s ?? throw new NullReferenceException(("stream null")));
+                }
+            }
+            private set => File.WriteAllText(VersionPath(), value.ToString(CultureInfo.InvariantCulture));
+        }
+
+        /// <summary>
+        ///   Upgrade/downgrade to the specified version.
+        /// </summary>
+        /// <param name="version">
+        ///   The required version of the repository.
+        /// </param>
+        /// <param name="cancel">
+        /// </param>
+        /// <returns></returns>
+        public async Task MirgrateToVersionAsync(int version, CancellationToken cancel = default)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The Cyclomatic Complexity of this method is 11 which is greater than 10 authorized.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)